Vald architecture document (#366)
vdaas/vald: docs/overview/architecture.md

  - [Search](#search)
- [Components](#components)
  - [Vald Filter](#vald-filter)
    - [Vald Ingress Filter](#vald-ingress-filter)
    - [Vald Egress Filter](#vald-egress-filter)
    - [Vald Filter Gateway](#vald-filter-gateway)
  - [Vald Metadata](#vald-metadata)
    - [Vald Meta Gateway](#vald-meta-gateway)
    - [Vald Meta](#vald-meta)
  - [Vald Backup](#vald-backup)
    - [Vald Compressor](#vald-compressor)
    - [Vald Backup Manager](#vald-backup-manager)
    - [Vald Backup Gateway](#vald-backup-gateway)
  - [Vald Load Balancing](#vald-load-balancing)
    - [Vald LB Gateway](#vald-lb-gateway)
    - [Agent Discoverer](#agent-discoverer)
  - [Vald Core Engine](#vald-core-engine)
    - [Vald Agent](#vald-agent)
    - [Vald Agent Scheduler](#vald-agent-scheduler)
    - [Vald Index Manager](#vald-index-manager)
  - [Vald Replication Manager](#vald-replication-manager)
    - [Vald Replication Manager Agent](#vald-replication-manager-agent)
    - [Vald Replication Manager Controller](#vald-replication-manager-controller)
  - [Kubernetes Components](#kubernetes-components)
    - [Kube-apiserver](#kube-apiserver)
    - [Custom Resources](#custom-resources)

## Overview

> Vald is based on [Kubernetes](https://kubernetes.io/) architecture.

Author: ```suggestion
Vald uses a cloud-native architecture focusing on [Kubernetes](https://kubernetes.io/).
Some components in Vald use Kubernetes API to control the behavior of distributed vector indexes.
```

Before you read this document, you must understand the basic concept of Kubernetes.

The below image is Vald's architecture.

<img src="../../design/Vald Future Architecture Overview.svg" />

We will explain this image in the following section.

## Data Flow

> In this section, we will explain the data flow in Vald which is the most important for users.

Reviewer: ```suggestion
This section describes the flow of data inside Vald and how Vald's vector indexes are stored.
This is the most important part of the user's understanding of Vald
```

Author: ```suggestion
This section describes the data flow inside Vald and how Vald's vector indexes are stored.
This is the most important part for the users to understand Vald.
```

### Insert

<img src="../../assets/docs/insert_flow.svg" />

When the user inserts data into Vald:

1. Vald Ingress receives the request from the user. The request includes the vector and the vector ID which is set by user.

> 2. Vald Ingress will forward the request to Vald Meta Gateway. After the Vald Meta Gateway receive the request, the UUID will be generated for internal use.

Reviewer: ```suggestion
2. Vald Ingress will forward the request to Vald Meta Gateway. After the Vald Meta Gateway receives the request, the UUID will be generated for internal use and mapped it with vector ID from the user's request.
```

3. Vald Meta Gateway will forward the request with UUID to Vald Backup Gateway.
4. Vald Backup Gateway will forward the request to Vald LB Gateway.

> 5. Vald LB Gateway will decide which Vald Agent instance to process the request based on the node resource usage and forward the request to the decided Vald Agent.

Reviewer: ```suggestion
5. Vald LB Gateway determines the number of Vald agent instances equal to the number of vector replicas to process the request based on the resource usage of the nodes and pods and transfers the request to the target Vald agent in parallel.
```

> 6. Vald Agent returns success and the corresponding Vald Agent IP address to Vald LB Gateway.

Reviewer: ```suggestion
6. Vald Agent stores the vectors and UUIDs in an on-memory vector queue and returns the success to the Vald LB gateway. A vector queue will be committed to an ANN graph index by a CreateIndex instruction executed by the index manager.
```

> 7. Vald LB Gateway returns success and the IP address to Vald Backup Gateway.

Reviewer: ```suggestion
7. Vald LB Gateway receives a success from each Replica's Agent and responds to the Backup Gateway with the IP address of each successful Agent.
```

Author: ```suggestion
7. Vald LB Gateway receives a success result from each Replica's Agent and responds to the Backup Gateway with the IP address of each successful Agent.
```

8. Vald Backup Gateway returns success to Vald Meta Gateway.

> 9. Vald Meta Gateway will send the metadata including the UUID and the vector ID to the Vald Meta.

Reviewer: ```suggestion
9. Vald Meta Gateway stores metadata such as UUIDs and vector IDs that were successfully inserted into the Agent in Vald Meta.
```

> 10. Vald Meta will store the metadata to the persistent layer.

Reviewer: ```suggestion
10. Vald Meta will store the metadata to the persistent layer such as Redis, Cassandra, MySQL, etc...
```

> 11. Vald Backup Gateway will send all the data (vector, vector ID and UUID) the Vald Compressor.

Reviewer: ```suggestion
11. Vald Backup Gateway will send all inserted the data (vector, vector ID and UUID)  to the Vald Compressor.
```

> 12. Vald Compressor will compress the data received and send to the Vald Backup Manager.

Reviewer: ```suggestion
12. Vald Compressor compresses data asynchronously and transfers the compressed data to Vald Backup Manager to reduce persistent storage usage.
```

Author: ```suggestion
12. Vald Compressor compresses data asynchronously to reduce persistent storage usage and transfers the compressed data to Vald Backup Manager.
```

> 13. Vald Backup Manager will store the compressed data to the persistent layer.

Reviewer: ```suggestion
13. Vald Backup Manager will store the compressed data to the persistent layer such as MySQL, Cassandra, etc...
```

### Search

<img src="../../assets/docs/search_flow.svg" />

When the user searches a vector from Vald:

> 1. Vald Ingress receives the search request from the user. The request includes the vector or the vector ID.

Reviewer: ```suggestion
1. Vald Ingress receives a search request from the user, Vald has a Search interface to search by vector and a SearchByID interface to search by stored vector index IDs, so that the request can be searched using a vector or vector ID.
```

# Vald Architecture <!-- omit in toc -->

> 

Reviewer 2: I think the description of this article should be top of article.
```suggestion
This document describes the high-level architecture design of Vald and explains each component in Vald.
```

  - [Search](#search)
- [Components](#components)
  - [Vald Filter](#vald-filter)
    - [Vald Ingress Filter](#vald-ingress-filter)
    - [Vald Egress Filter](#vald-egress-filter)
    - [Vald Filter Gateway](#vald-filter-gateway)
  - [Vald Metadata](#vald-metadata)
    - [Vald Meta Gateway](#vald-meta-gateway)
    - [Vald Meta](#vald-meta)
  - [Vald Backup](#vald-backup)
    - [Vald Compressor](#vald-compressor)
    - [Vald Backup Manager](#vald-backup-manager)
    - [Vald Backup Gateway](#vald-backup-gateway)
  - [Vald Load Balancing](#vald-load-balancing)
    - [Vald LB Gateway](#vald-lb-gateway)
    - [Agent Discoverer](#agent-discoverer)
  - [Vald Core Engine](#vald-core-engine)
    - [Vald Agent](#vald-agent)
    - [Vald Agent Scheduler](#vald-agent-scheduler)
    - [Vald Index Manager](#vald-index-manager)
  - [Vald Replication Manager](#vald-replication-manager)
    - [Vald Replication Manager Agent](#vald-replication-manager-agent)
    - [Vald Replication Manager Controller](#vald-replication-manager-controller)
  - [Kubernetes Components](#kubernetes-components)
    - [Kube-apiserver](#kube-apiserver)
    - [Custom Resources](#custom-resources)

## Overview

> This document describes the high-level architecture design of Vald and explains each component in Vald.

Reviewer 2: ```suggestion

```

> We need these components to support scalability, high performance, and auto-healing in Vald.

Reviewer 2: I'm not sure, this sentence is suited here 😅

Author: If we remove the sentence before, it is not suit here..

Reviewer 2: yes.
Do you have any ideas?

Author: I am thinking it is better to remove the sentence, what do you think?

> <img src="../../design/Vald Future Architecture Overview.svg" />

Reviewer 2: ```suggestion
```

> Vald is based on [Kubernetes](https://kubernetes.io/) architecture. Before you read this document, you must understand the basic concept of Kubernetes.

Reviewer 2: ```suggestion
Vald is based on [Kubernetes](https://kubernetes.io/) architecture.
Before you read this document, you must understand the basic concept of Kubernetes.
The below image is Vald's architecture.

<img src="../../design/Vald Future Architecture Overview.svg" />

We mention this image following sections.
```

> Vald is based on [Kubernetes](https://kubernetes.io/) architecture. Before you read this document, you must understand the basic concept of Kubernetes.

Author: How about this?

```suggestion
Vald is based on [Kubernetes](https://kubernetes.io/) architecture.
Before you read this document, you must understand the basic concept of Kubernetes.

The below image is Vald's architecture.

<img src="../../design/Vald Future Architecture Overview.svg" />

We will explain this image in the following section.
```

Reviewer 2: LGTM

## Data Flow

In this section, we will explain the data flow in Vald which is the most important for users.

### Insert

<img src="../../assets/docs/insert_flow.svg" />

When the user inserts data into Vald:

> ```bash
> 1. Vald Ingress receives the request from the user. The request includes the vector and the vector ID which is set by user.
> 2. Vald Ingress will forward the request to Vald Meta Gateway. After the Vald Meta Gateway receive the request, the UUID will be generated for internal use.
> 3. Vald Meta Gateway will forward the request with UUID to Vald Backup Gateway.
> 4. Vald Backup Gateway will forward the request to Vald LB Gateway.
> 5. Vald LB Gateway will decide which Vald Agent instance to process the request based on the node resource usage and forward the request to the decided Vald Agent.
> 6. Vald Agent returns success and the corresponding Vald Agent IP address to Vald LB Gateway.
> 7. Vald LB Gateway returns success and the IP address to Vald Backup Gateway.
> 8. Vald Backup Gateway returns success to Vald Meta Gateway.
> 9. Vald Meta Gateway will send the metadata including the UUID and the vector ID to the Vald Meta.
> 10. Vald Meta will store the metadata to the persistent layer.
> 11. Vald Backup Gateway will send all the data (vector, vector ID and UUID) the Vald Compressor.
> 12. Vald Compressor will compress the data received and send to the Vald Backup Manager.
> 13. Vald Backup Manager will store the compressed data to the persistent layer.
> ```

Reviewer 3: ![badge](https://img.shields.io/badge/review-ask-blueviolet.svg?style=flat-square)
Why is this using code block and bash syntax highlight?

i think it's better to use numbered list simply.

```suggestion
1. Vald Ingress receives the request from the user. The request includes the vector and the vector ID which is set by user.
2. Vald Ingress will forward the request to Vald Meta Gateway. After the Vald Meta Gateway receive the request, the UUID will be generated for internal use.
3. Vald Meta Gateway will forward the request with UUID to Vald Backup Gateway.
4. Vald Backup Gateway will forward the request to Vald LB Gateway.
5. Vald LB Gateway will decide which Vald Agent instance to process the request based on the node resource usage and forward the request to the decided Vald Agent.
6. Vald Agent returns success and the corresponding Vald Agent IP address to Vald LB Gateway.
7. Vald LB Gateway returns success and the IP address to Vald Backup Gateway.
8. Vald Backup Gateway returns success to Vald Meta Gateway.
9. Vald Meta Gateway will send the metadata including the UUID and the vector ID to the Vald Meta.
10. Vald Meta will store the metadata to the persistent layer.
11. Vald Backup Gateway will send all the data (vector, vector ID and UUID) the Vald Compressor.
12. Vald Compressor will compress the data received and send to the Vald Backup Manager.
13. Vald Backup Manager will store the compressed data to the persistent layer.
```

Author: I intended to force user to focus on this block... Do you think we need the coding block?

Reviewer 3: I see.

I think we don't need to use code block here.
and i think it's not a good UX because the long line will not be wrapped and users should scroll horizontally to read them.
<img width="1003" alt="スクリーンショット 2020-05-14 18 37 48" src="https://user-images.githubusercontent.com/1588935/81918895-112d9880-9612-11ea-9a52-bad19a9e97a0.png">

### Search

<img src="../../assets/docs/search_flow.svg" />

When the user searches a vector from Vald:

> ```bash
> 1. Vald Ingress receives the search request from the user. The request includes the vector or the vector ID.
> 2. Vald Ingress will forward the request to Vald Filter Gateway.
> 3. Vald Filter Gateway will forward the request to Vald Ingress Filter.
> 4. Vald Ingress Filter will perform the filtering and return the filtering result to the Vald Filter Gateway.
> 5. Vald Filter Gateway will forward the request to the Vald Meta Gateway.
> 6. Vald Meta Gateway will forward the request to the Vald Backup Gateway.
> 7. Vald Backup Gateway will forward the request to the Vald LB Gateway.
> 8. Vald LB Gateway will decide which Vald Agent instance to process the request base on the node resource usage and forwards the request to the decided Vald Agent.
> 9. Vald Agent returns the searching result to the Vald LB Gateway. The searching result includes the UUID.
> 10. Vald LB Gateway returns the searching result to Vald Backup Gateway.
> 11. Vald Backup Gateway returns the searching result to the Vald Meta Gateway.
> 12. Vald Meta Gateway will forward the searching result to the Vald Meta.
> 13. Vald Meta will perform a search for the Vector ID base on the UUID and return the Vector ID to the Vald Meta Gateway.
> 14. Vald Meta Gateway returns the searching result with the vector ID to the Vald Filter Gateway.
> 15. Vald Filter Gateway will forward the request to Vald Egress Gateway to filter the final result.
> ```

Reviewer 3: same.

```suggestion
1. Vald Ingress receives the search request from the user. The request includes the vector or the vector ID.
2. Vald Ingress will forward the request to Vald Filter Gateway.
3. Vald Filter Gateway will forward the request to Vald Ingress Filter.
4. Vald Ingress Filter will perform the filtering and return the filtering result to the Vald Filter Gateway.
5. Vald Filter Gateway will forward the request to the Vald Meta Gateway.
6. Vald Meta Gateway will forward the request to the Vald Backup Gateway.
7. Vald Backup Gateway will forward the request to the Vald LB Gateway.
8. Vald LB Gateway will decide which Vald Agent instance to process the request base on the node resource usage and forwards the request to the decided Vald Agent.
9. Vald Agent returns the searching result to the Vald LB Gateway. The searching result includes the UUID.
10. Vald LB Gateway returns the searching result to Vald Backup Gateway.
11. Vald Backup Gateway returns the searching result to the Vald Meta Gateway.
12. Vald Meta Gateway will forward the searching result to the Vald Meta.
13. Vald Meta will perform a search for the Vector ID base on the UUID and return the Vector ID to the Vald Meta Gateway.
14. Vald Meta Gateway returns the searching result with the vector ID to the Vald Filter Gateway.
15. Vald Filter Gateway will forward the request to Vald Egress Gateway to filter the final result.
```

Vald Ingress Filter filters the incoming request before processing it.

Users can implement custom filtering logic such as changing the vectors or filtering based on user ID.

#### Vald Egress Filter

Vald Egress Filter filters the response before sending it to the user.

This component can reorder the searching result from multiple Vald Agents based on the user-defined ranking.

#### Vald Filter Gateway

Vald Filter Gateway forwards the request to Vald Ingress Filter before processing it and forwards the response to the Vald Egress Filter before returning the searching result to the user.

### Vald Metadata

In Vald, metadata consists of the vector data and the corresponding additional data to represent the set of the searching criteria and the result.

Vald Metadata includes the user inputted metadata(vector ID) and the vector, and the internal generated UUID.

#### Vald Meta Gateway

The main responsibility of the Vald Meta Gateway is to process the Vald metadata and to forward the information to Vald Backup Gateway.

It will perform the following action:

1. Return error if the user has already input the same vector in Vald
1. Generate the corresponding UUID for internal use.

> 1. Forward the metadata (vector ID and UUID) request to the Vald Meta Agent.

Author: I think it is more clear, what do you think?
```suggestion
1. Forward the vector ID and UUID request to the Vald Meta.
```

Reviewer 3: I think it is good. :+1:

1. Forward the vector information (vector ID, vector, and UUID) to Vald Backup Gateway.

#### Vald Meta

Vald Meta is the agent to process the CRUD request of the metadata (vector ID and UUID).
Users can configure which data source to be used in Vald Meta (for example Redis or Cassandra).

### Vald Backup

To support auto-healing functionality and increase performance during disaster recovery, Vald implements the backup mechanism.

#### Vald Compressor

> Vald Compressor compresses all of the data (metadata and the vector data) and sends to the Vald Backup Manager to process the backup request.

Author: ```suggestion
Vald Compressor compresses the vector data and sends to the Vald Backup Manager to process the backup request.
```

#### Vald Backup Manager

> Vald Backup Manager processes the CRD request of the backup request and handles the compressed metadata. Users can configure which data source to be used in Vald Meta (for example Redis or Cassandra).

Author: ```suggestion
Vald Backup Manager processes the Create/Read/Delete request of the backup request and handles the compressed metadata. Users can configure which data source to be used in Vald Meta (for example Redis or Cassandra).
```

#### Vald Backup Gateway

Vald Backup Gateway will forward the backup request to the Vald LB Gateway.
It also forwards to Vald Compressor asynchronously with metadata.

### Vald Load Balancing

Load balancing is one of the important concepts in distributed computing, which means it distributes a set of tasks over a set of resources aiming for making the overall processing more efficient.
Vald implements its own load balancing controller.
Vald can load balance the request base on node resources.

#### Vald LB Gateway

Vald LB Gateway loads balance the user request base on the node resources results from the Agent Discoverer.

#### Agent Discoverer

Agent Discoverer discovers active Vald pods and the corresponding node's resources usage via [kube-apiserver](https://github.com/kubernetes/kubernetes/tree/master/cmd/kube-apiserver).

### Vald Core Engine

In this section, we will describe what is Vald Agent and the corresponding components to support Vald Agent.

#### Vald Agent

> Vald Agent is the core of the Vald. By default Vald uses [NGT](https://github.com/yahoojapan/NGT) to provide API for users to insert/update/delete/search vectors.

Reviewer 3: ![badge](https://img.shields.io/badge/review-suggestion-blue.svg?style=flat-square)
```suggestion
Vald Agent provides functionalities to perform approximate nearest neighbor search.
Agent-NGT uses [yahoojapan/NGT](https://github.com/yahoojapan/NGT) as a core library.
```

> Each Vald Agent pod holds different high dimensional vector data space, which is constructed by insert/update vectors for searching approximate vectors.
> 
> When you request searching with your vector in Vald, each Vald Agent returns different _k_-nearest neighbors' vectors which are similar to the searching vector.
> 
> <img src="../../assets/docs/vector_data_space_explain.svg" />
> 
> The same vector will be inserted into multiple Vald Agents, not all Vald Agents.
> It makes that the vector data space constructed in each Vald Agent will be different, causing the searching results from each Vald Agent are different.

Reviewer 3: ![badge](https://img.shields.io/badge/review-suggestion-blue.svg?style=flat-square)

```suggestion
Each Vald Agent pod has own vector data space because only several pods are selected to be inserted every single vector.
When you send a search request to Vald, each Vald Agent return different results of _k_-nearest neighbors depends on their index, and you'll get the merged result of them.

<img src="../../assets/docs/vector_data_space_explain.svg" />
```

Reviewer 3: i'm not sure it is valid syntax, but it can be more simpler than before.

Author: I think it is easier to understand, thank you for your suggestion :)
just corrected some wordings, please review :).
```suggestion
Each Vald Agent pod has its own vector data space because only several Vald Agents are selected to be inserted/updated in a single insert/update request.

When searching a vector in Vald, each Vald Agent return different results of _k_-nearest neighbors depending on their index, and you'll get the merged result of them.
<img src="../../assets/docs/vector_data_space_explain.svg" />
```

Reviewer 3: Great! :+1:

    - [Vald Replication Manager Controller](#vald-replication-manager-controller)
  - [Kubernetes Components](#kubernetes-components)
    - [Kube-apiserver](#kube-apiserver)
    - [Custom Resources](#custom-resources)

## Overview

Vald uses a cloud-native architecture focusing on [Kubernetes](https://kubernetes.io/).
Some components in Vald use Kubernetes API to control the behavior of distributed vector indexes.
Before reading this document, you need to have some understanding of the basic idea of cloud-native architecture and Kubernetes.

The below image is Vald's architecture.

<img src="../../design/Vald Future Architecture Overview.svg" />

We will explain this image in the following section.

## Data Flow

This section describes the data flow inside Vald and how Vald's vector indexes are stored.
This is the most important part for the users to understand Vald.

### Insert

<img src="../../assets/docs/insert_flow.svg" />

When the user inserts data into Vald:

1. Vald Ingress receives the request from the user. The request includes the vector and the vector ID.
2. Vald Ingress will forward the request to the Vald Filter Gateway to pre-process the request data.

> 3. Vald Filter Gateway will forward the request to the user-defined Vald Ingress Filter. After the Vald Ingress Filter received the request, it will perform the pre-processing logic defined by the user, for example, padding the vector to match the vector length in Vald.

Reviewer 2: ```suggestion
3. Vald Filter Gateway will forward the request to the user-defined Vald Ingress Filter. After the Vald Ingress Filter received the request, it will perform the pre-processing logic defined by the user, for example, padding the vector to match the vector dimensionality in Vald.
```

4. After the request is processed by the user-defined Vald Ingress Filter, the result will return to the Vald Filter Gateway.

> 5. Vald Filter Gateway will forward the processed data to the Vald Meta Gateway. Vald Meta Gateway will generate the UUID for internal use and the UUID will be mapped to the vector ID from the user's request. The reason of using UUID instead of vector ID is because the vector ID may be too long and it may increase the memory usage in Vald Agent.

Reviewer 2: ```suggestion
5. Vald Filter Gateway will forward the processed data to the Vald Meta Gateway. Vald Meta Gateway will generate the UUID each requested vector for internal use and the UUID will be mapped to the vector ID from the user's request. The reason of using UUID instead of vector ID is because the vector ID may be too long and it may increase the memory usage in Vald Agent.
```

6. Vald Meta Gateway will forward the request with the UUID to the Vald Backup Gateway, which will process the backup logic in 14-16 to prevent the data lost in Vald.
7. Vald Backup Gateway will forward the request to Vald LB Gateway. Vald LB Gateway will determine which Vald Agent(s) to process the request based on the resource usage of the nodes and pods, and the number of vector replicas.

> 8. Vald LB Gateway will forward the UUID and the vector data to the selected Vald Agents parallelly. Vald Agent will insert the vector and UUID in an on-memory vector queue. A vector queue will be committed to an ANN graph index by a `CreateIndex` instruction executed by the Vald Index Manager.

Reviewer 2: ```suggestion
8. Vald LB Gateway will forward the UUID and the vector data to the selected Vald Agents in parallel. Vald Agent will insert the vector and UUID in an on-memory vector queue. A vector queue will be committed to an ANN graph index by a `CreateIndex` instruction executed by the Vald Index Manager.
```

### Search

<img src="../../assets/docs/search_flow.svg" />

When the user searches a vector from Vald:

1. Vald Ingress receives a search request from the user. Vald provides 2 searching interfaces to the user, the user can search by vector or search by the vector ID.
2. Vald Ingress will forward the request to the Vald Filter Gateway to pre-process the request data.
3. Vald Filter Gateway will forward the request to the user-defined Vald Ingress Filter. After the Vald Ingress Filter received the request, it will perform the pre-processing logic defined by the user, for example, padding the vector to match the vector dimension in Vald.
4. After the request is processed by the user-defined Vald Ingress Filter, the result will return to the Vald Filter Gateway.
5. Vald Filter Gateway will forward the request to the Vald Meta Gateway. Vald Meta Gateway is used to resolve the internal used UUID to the user inserted vector ID in step 10-11.
6. Vald Meta Gateway will forward the request to the Vald LB Gateway. Vald LB Gateway will preform the post-processing of the result in step 9 after the Vald Agent(s) return in step 8.

> 7. Vald LB Gateway will forward the request to all Vald Agents in parallel. Each Vald Agent will search the _k_ nearest neighbor vectors in an on memory cache.

Reviewer: ```suggestion
7. Vald LB Gateway will forward the request to all Vald Agents in parallel. Each Vald Agent will search the _k_ nearest neighbor vectors in an on memory graph index.
```

8. Vald Agent returns the searching result to the Vald LB Gateway. The searching result includes the UUID, the vector distance, and the vector. The number of the result will be the same as requested.

> 9. After all Vald Agents return the searching result, Vald LB Gateway will aggregate all of the results and reorder the result by the vector distance, and return the reordered result to the Vald Meta Gateway.

Reviewer 2: - I'm not sure `reorder` is the most suited word in this sentence... Where to Vald LB Gateway reorder?

Reviewer 4: I think reorder means relank search result vector:thinking:

Reviewer 2: I see your consideration, and I think `reoreder` is not suit and not correct in this sentence.
User may think `reoreder` means "Vald Meta Gateway request to Vald Agents again in 9".

Author: ```suggestion
9. Vald LB Gateway will aggregate all searching results from all Vald Agents, rank the result by the vector distance, and return the ranked result to the Vald Meta Gateway.
```

Reviewer 4: Thank you for your changes.
I think it is good.

10. Vald Meta Gateway will forward the searching result to the Vald Meta to resolve the user-defined vector IDs from the internal used UUIDs.

> 11. Vald Meta will perform a search for the Vector IDs base on the internal used UUIDs.

Reviewer 2: ```suggestion
11. Vald Meta will perform a search for the Vector IDs based on the internal used UUIDs.
```

#### Vald Egress Filter

Vald Egress Filter filters the response before sending it to the user.

This component can reorder the searching result from multiple Vald Agents based on the user-defined ranking.

#### Vald Filter Gateway

Vald Filter Gateway forwards the request to Vald Ingress Filter before processing it and forwards the response to the Vald Egress Filter before returning the searching result to the user.

### Vald Metadata

In Vald, metadata consists of the vector data and the corresponding additional data to represent the set of the searching criteria and the result.

Vald Metadata includes the user inputted metadata(vector ID) and the vector, and the internal generated UUID.

#### Vald Meta Gateway

The main responsibility of the Vald Meta Gateway is to process the Vald metadata and to forward the information to Vald Backup Gateway.

It will perform the following action:

1. Return error if the user has already input the same vector in Vald
1. Generate the corresponding UUID for internal use.
1. Forward the vector ID and UUID request to the Vald Meta.
1. Forward the vector information (vector ID, vector, and UUID) to Vald Backup Gateway.

> 

Reviewer 5: why are all numbers 1?

Reviewer 2: It is github markdown coding style.
1. a
1. b
1. c
